src/sbvr-api/uri-parser.ts

```diff
@@ -0,0 +1,225 @@
+import * as _AbstractSQLCompiler from '@resin/abstract-sql-compiler'
+
+import * as Promise from 'bluebird'
+import TypedError = require('typed-error')
+const { ODataParser } = require('@resin/odata-parser')
+interface OData2AbstractSQLInstance {
+	match: (abstractSQL: ODataQuery, rule: 'Process', args: [SupportedMethod, string[]]) => {
+		tree: _AbstractSQLCompiler.AbstractSqlQuery,
+		extraBodyVars: {},
+	}
+	setClientModel: (abstractSqlModel: _AbstractSQLCompiler.AbstractSqlModel) => void
+}
+const { OData2AbstractSQL }: {
+	OData2AbstractSQL: {
+		createInstance: () => OData2AbstractSQLInstance
+	}
+} = require('@resin/odata-to-abstract-sql')
+import * as memoize from 'memoizee'
+import * as _ from 'lodash'
+
+export class TranslationError extends TypedError {}
+export class ParsingError extends TypedError {}
+export class BadRequestError extends TypedError {}
+
+type SupportedMethod = "GET" | "PUT" | "POST" | "PATCH" | "MERGE" | "DELETE" | "OPTIONS"
+type ODataQuery = any[]
+
+export interface UnparsedRequest {
+	method: string
+	url: string
+	data: any
+	headers?: { [header: string]: string }
+	changeSet?: UnparsedRequest[]
+	_isChangeSet?: boolean
+}
+
+export interface ODataRequest {
+	method: SupportedMethod
+	odataQuery: ODataQuery
+	odataBinds: any[]
+	values: any[]
+	abstractSqlQuery?: _AbstractSQLCompiler.AbstractSqlQuery | _AbstractSQLCompiler.AbstractSqlQuery[]
+	sqlQuery?: _AbstractSQLCompiler.SqlResult | _AbstractSQLCompiler.SqlResult[]
+	resourceName: string
+	vocabulary: string
+	_defer?: boolean
+	id?: number
+	custom?: {
+		[ key: string ]: any
+	}
+}
+
+const odataParser = ODataParser.createInstance()
+const odata2AbstractSQL: {
+	[vocabulary: string]: OData2AbstractSQLInstance
+} = {}
+
+// Converts a value to its string representation and tries to parse is as an
+// OData bind
+export const parseId = (b: any) => {
+	return ODataParser.matchAll(String(b), 'ExternalKeyBind')
+}
+
+const memoizedOdata2AbstractSQL = (() => {
+	const memoizedOdata2AbstractSQL = memoize(
+		(vocabulary: string, odataQuery: ODataQuery, method: SupportedMethod, bodyKeys: string[]) => {
+			try {
+				return odata2AbstractSQL[vocabulary].match(odataQuery, 'Process', [method, bodyKeys])
+			} catch (e) {
+				console.error('Failed to translate url: ', JSON.stringify(odataQuery, null, '\t'), method, e, e.stack)
+				throw new TranslationError('Failed to translate url')
+			}
+		},
+		{ normalizer: JSON.stringify }
+	)
+	return (vocabulary: string, odataQuery: ODataQuery, method: SupportedMethod, body: {}) => {
+		// Sort the body keys to improve cache hits
+		const { tree, extraBodyVars } = memoizedOdata2AbstractSQL(vocabulary, odataQuery, method, _.keys(body).sort())
+		_.assign(body, extraBodyVars)
```

Comment:
I know this is pre-existing, but does this function return a value *and* mutate the `body` parameter? Is there a cleaner way of doing this?